crates/core/src/rlp/decode.rs

```diff
@@ -0,0 +1,454 @@
+}
+
+impl RLPDecode for ethereum_types::H264 {
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        RLPDecode::decode(rlp).map(ethereum_types::H264)
+    }
+}
+
+impl RLPDecode for ethereum_types::Address {
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        RLPDecode::decode(rlp).map(ethereum_types::H160)
+    }
+}
+
+impl RLPDecode for ethereum_types::H512 {
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        RLPDecode::decode(rlp).map(ethereum_types::H512)
+    }
+}
+
+impl RLPDecode for ethereum_types::Signature {
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        RLPDecode::decode(rlp).map(ethereum_types::H520)
+    }
+}
+
+impl RLPDecode for String {
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        let str_bytes = decode_bytes(rlp)?.0.to_vec();
+        String::from_utf8(str_bytes).map_err(|_| RLPDecodeError::MalformedData)
+    }
+}
+
+impl RLPDecode for Ipv4Addr {
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        let (ip_bytes, _) = decode_bytes(rlp)?;
+        let octets: [u8; 4] = ip_bytes
+            .try_into()
+            .map_err(|_| RLPDecodeError::InvalidLength)?;
+        Ok(Ipv4Addr::from(octets))
+    }
+}
+
+impl RLPDecode for Ipv6Addr {
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        let (ip_bytes, _) = decode_bytes(rlp)?;
+        let octets: [u8; 16] = ip_bytes
+            .try_into()
+            .map_err(|_| RLPDecodeError::InvalidLength)?;
+        Ok(Ipv6Addr::from(octets))
+    }
+}
+
+impl RLPDecode for IpAddr {
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        let (ip_bytes, _) = decode_bytes(rlp)?;
+
+        match ip_bytes.len() {
+            4 => {
+                let octets: [u8; 4] = ip_bytes
+                    .try_into()
+                    .map_err(|_| RLPDecodeError::InvalidLength)?;
+                Ok(IpAddr::V4(Ipv4Addr::from(octets)))
+            }
+            16 => {
+                let octets: [u8; 16] = ip_bytes
+                    .try_into()
+                    .map_err(|_| RLPDecodeError::InvalidLength)?;
+                Ok(IpAddr::V6(Ipv6Addr::from(octets)))
+            }
+            _ => Err(RLPDecodeError::InvalidLength),
+        }
+    }
+}
+
+impl<T: RLPDecode> RLPDecode for Vec<T> {
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        // empty RLP encoded list must have at least the RLP_EMPTY_LIST byte
+        if rlp.is_empty() {
+            return Err(RLPDecodeError::InvalidLength);
+        }
+
+        // empty list case, [ 0xc0 ]
+        if rlp[0] == RLP_EMPTY_LIST {
+            return Ok(Vec::new());
+        }
+
+        // extract the list length and return the payload
+        let payload = decode_list(rlp)?;
+        let mut result = Vec::new();
+        let mut current_slice = payload;
+
+        while !current_slice.is_empty() {
+            let (_, rest) = decode_bytes(current_slice)?;
```

Comment:
I don't think this will work correctly for `Vec<Vec<T>>`. Can we add tests for that?

Reply:
added!